### PR TITLE
Allow use of prime symbol in type names.

### LINKED ISSUE #1569: Allow prime symbols in types
I don't think that it's very hard to implement (maybe I'm wrong), but I think it can be useful sometimes. Here is an example:

```purescript
module Foo_Test where

import Test.QuickCheck
import Test.QuickCheck.Arbitrary
import Foo (Foo(..))

newtype Foo' = Foo' Foo

instance arbitraryFoo' :: Arbitrary Foo' where ...
```

Underscore could be useful, for example, for some tool that automatically finds all test modules (or something else) and runs them and prime symbol could be useful when you have some wrapper, but you don't want to create a meaningful name for it.

### PR DESCRIPTION
**Pull request by garyb**
Resolves #1569

### CODE REVIEW

**Comment by garyb** [comment]
I opened the last PR from the GitHub client, either forgot to switch the branch or something odd happened :smile:

File: src/Language/PureScript/Parser/Lexer.hs [hunk 1]
```diff
@@ -213,8 +216,10 @@ parseToken = P.choice
   , P.try $ P.char '_'    *> P.notFollowedBy identLetter *> pure Underscore
   , HoleLit <$> P.try (P.char '?' *> P.many1 identLetter)
   , LName         <$> parseLName
-  , do uName <- parseUName
-       (guard (validModuleName uName) >> Qualifier uName <$ P.char '.') <|> pure (UName uName)
+  , parseUName >>= \uName ->
+      (guard (validModuleName uName) >> Qualifier uName <$ P.char '.')
+        <|> (guard (validUName uName) >> pure (UName uName))
```
Comment (paf31) [inline src/Language/PureScript/Parser/Lexer.hs]: Does it make sense to move these checks into `mname`, `uname`, etc.

Comment (garyb) [inline src/Language/PureScript/Parser/Lexer.hs]: That sounds good to me, I certainly found this a bit strange. Shall I do that then?

**Comment by garyb** [comment]
I need to update the parsing of `DeclarationRef` to use `typeName` parsing too. I'll push that with any other revisions in a moment.

File: src/Language/PureScript/Parser/Lexer.hs (hunk 1 above)
Comment (garyb) [inline src/Language/PureScript/Parser/Lexer.hs]: Actually, can we? The `P.char '.'` thing in there for `Qualifier` probably obstructs that. Should work for `uname` / `tyname` though.

**Comment by garyb** [comment]
`DeclarationRef` handling and lexer tweak now in.

File: src/Language/PureScript/Parser/Lexer.hs [hunk 2]
```diff
@@ -440,12 +444,6 @@ mname = token go P.<?> "module name"
   go (UName s) | validModuleName s = Just s
   go _ = Nothing
 
-uname' :: String -> TokenParser ()
-uname' s = token go P.<?> show s
-  where
-  go (UName s') | s == s' = Just ()
-  go _ = Nothing
```
Comment (garyb) [inline src/Language/PureScript/Parser/Lexer.hs]: Deleted this as it was unused, and I wasn't sure what the right implementation would be anymore: should the `validUName` predicate apply? etc.

**Comment by paf31** [comment]
🚢 thanks!

File: src/Language/PureScript/Parser/Lexer.hs (hunk 2 above)
Comment (paf31) [inline src/Language/PureScript/Parser/Lexer.hs]: If it's unused, just get rid, I think.

**Comment by garyb** [comment]
Looks like the nightly build is broken now? Not by this I imagine, there's some message about the compiler.

**Comment by paf31** [comment]
Yes, the Mac build has been broken for a few days. I've been ignoring it, but maybe we should turn that build off.

**Pull request merged**